source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,340 @@
+  // Step 2: Verify each response header
+  //
+  std::string reject_details;
+  header_map.iterate([this, &reject_details](const ::Envoy::Http::HeaderEntry& header_entry)
+                         -> ::Envoy::Http::HeaderMap::Iterate {
+    const auto& header_name = header_entry.key();
+    const auto& header_value = header_entry.value();
+    const auto& string_header_name = header_name.getStringView();
+
+    if (string_header_name.empty()) {
+      reject_details = UhvResponseCodeDetail::get().EmptyHeaderName;
+    } else if (string_header_name.at(0) == ':' &&
+               !kAllowedPseudoHeaders.contains(header_name.getStringView())) {
+      // This is an unrecognized pseudo header, reject the response
+      reject_details = UhvResponseCodeDetail::get().InvalidPseudoHeader;
+    } else {
+      auto entry_result = validateResponseHeaderEntry(header_name, header_value);
+      if (!entry_result) {
+        reject_details = static_cast<std::string>(entry_result.details());
+      }
+    }
+
+    return reject_details.empty() ? ::Envoy::Http::HeaderMap::Iterate::Continue
+                                  : ::Envoy::Http::HeaderMap::Iterate::Break;
+  });
+
+  return reject_details.empty()
+             ? ResponseHeaderMapValidationResult::success()
+             : ResponseHeaderMapValidationResult(RejectAction::Reject, reject_details);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateTEHeader(const ::Envoy::Http::HeaderString& value) {
+  //
+  // Only allow a TE value of "trailers" for HTTP/2, based on
+  // RFC 7540, https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2:
+  //
+  // The only exception to this is the TE header field, which MAY be present
+  // in an HTTP/2 request; when it is, it MUST NOT contain any value other
+  // than "trailers".
+  //
+  if (!absl::EqualsIgnoreCase(value.getStringView(), header_values_.TEValues.Trailers)) {
+    return {RejectAction::Reject, Http2ResponseCodeDetail::get().InvalidTE};
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateAuthorityHeader(const ::Envoy::Http::HeaderString& value) {
+  //
+  // From RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.2:
+  //
+  // authority = [ userinfo "@" ] host [ ":" port ]
+  //
+  // HTTP/2 deprecates the userinfo portion of the :authority header. Validate
+  // the :authority header and reject the value if the userinfo is present. This
+  // is based on RFC 7540, https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.3
+  //
+  // The host portion can be any valid URI host, which this function does not
+  // validate. The port, if present, is validated as a valid uint16_t port.
+  //
+  return validateHostHeader(value);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateGenericHeaderName(const ::Envoy::Http::HeaderString& key) {
```

Comment:
This should also check that the name is lower case, since `HeaderValidator::validateGenericHeaderName` allows both lower and upper case.